Static example tests

acwooding opened the pull request:
- Sets up initial static image tests using existing examples. 
- Sets up azure pipelines to run the tests and keep the image test artifacts. 
- Cleans up duplicated test report calls in the pipeline (since it's using pytest-azurepipelines already). 
- Adds a pip cache to the pipeline (fwiw since it only has test dependencies atm, but could easily include the main deps to potentially speed things up more). 
- Adds a scheduled weekly CI run on Sundays.

Note that two of the tests are currently marked as xfail if running in azure pipelines as the CI generated images differ by a pixel in size in one dimension and any size mismatch is an automatic fail (dealing with that will be future work and a separate PR).

gclendenning commented:
Overall I think it looks good :) It's probably worth adding a bit of documentation since there are a few moving pieces. If I'm understanding correctly the name of the script and the name of the image file in the baseline folder need to match and the actual comparison is done using the ```@pytest.mark.mpl_image_compare``` decorator. Plus we'll definitely need documentation for the interactive tests so wouldn't hurt to have it for the static ones as well. 

We probably also want to test on other OS/python versions. There are examples of doing this in the other pipelines ([example](https://github.com/lmcinnes/pynndescent/blob/master/azure-pipelines.yml))

acwooding commented:
Yes, pytest-mpl manages the generation of baseline images by running `pytest --mpl-generate-path=datamapplot/tests/baseline`.  A test function with the decorator @pytest.mark.mpl_image_compare has to return a figure and the rest is automated is taken care of by pytest-mpl as the names are automatically generated by the name of test function. 

I didn't write this up in documentation yet since it's more mechanics of how pytest-mpl works rather than somethign that goes in docstrings, and I wanted to check where to put the test instructions (in a README in the tests directory?). Is there a preference for where the documentation for how the tests work should live?

I left the other OS/python versions commented out for now because that's the way I found things and wasn't sure when it would be desirable to uncomment them.

acwooding commented:
I added some docstrings to the relevant functions now.

I updated to include the mac and linux versions for 3.10 and 3.11. Looks like 3.9 isn't supported because of: `from typing import Any, Callable, cast, ParamSpec, TypeVar, Union` since `ParamSpec` isn't in `typing` until python 3.10.

lmcinnes commented:
Thanks Amy, this looks great! With the infrastructure set up I can also add more tests as required. This also gives me more motivation to add more examples to the example gallery, since they can also be tests.
